File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.py
def merge_sort(a,s,e):
    if s < e:
        m = (s+e)//2
        merge_sort(a,s,m)
        merge_sort(a,m+1,e)
        merge(a,s,m,e)

def merge(a,s,m,e):
    global cnt, ans
    i, j, t = s, m+1, 0
    while i <= m and j <= e:
        if a[i] <= a[j]:
            tmp[t] = a[i]
            i += 1
        else:
            tmp[t] = a[j]
            j += 1
        t += 1
    while i <= m:
        tmp[t] = a[i]
        t += 1
        i += 1
    while j <= e:
        tmp[t] = a[j]
        t += 1
        j += 1
    i , t = s, 0    
    while i <= e:
        a[i] = tmp[t]
        cnt += 1
        if cnt == k:
            ans = a[i]
        i += 1
        t += 1
n, k = map(int, input().split())
a = list(map(int, input().split()))
tmp = [0]*n
cnt = 0
ans = 0
merge_sort(a,0,n-1)
if ans:
    print(ans)
else:
    print(-1)

File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.java
import java.io.IOException;

public class Main {
    static int[] tmp;
    static int cnt, ans, k;

    public static void main(String[] args) throws IOException {
        int n = readInt();
        cnt = 0;
        ans = -1;
        k = readInt();
        int[] a = new int[n];
        tmp = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = readInt();
        }

        merge_sort(a, 0, n-1);
        System.out.println(ans);
    }

    static void merge_sort(int[] a, int p, int r) {
        if (p < r) {
            int q = (p + r) / 2;
            merge_sort(a, p, q);
            merge_sort(a, q + 1, r);
            merge(a, p, q, r);
        }
    }

    static void merge(int[] a, int p, int q, int r) {
        int i = p, j = q + 1, t = 0;
        while (i <= q && j <= r) {
            if (a[i] <= a[j]) {
                tmp[t++] = a[i++];
            } else {
                tmp[t++] = a[j++];
            }
        }
        while (i <= q) {
            tmp[t++] = a[i++];
        }
        while (j <= r) {
            tmp[t++] = a[j++];
        }
        i = p;
        t = 0;
        while (i <= r) {
            cnt++;
            if (cnt == k) {
                ans = tmp[t];
            }
            a[i++] = tmp[t++];
        }
    }

    static int readInt() throws IOException {
        int v, t = 0;
        while ((v = System.in.read()) > 32) {
            t = t * 10 + v - 48;
        }
        if (v == 13) {
            System.in.read();
        }
        return t;
    }
}